### PR TITLE
Automatic `avg_num_neighbors`

### PR DESCRIPTION
**Pull request by Linux-cpp-lisp**
## How Has This Been Tested?
unit tests

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds or improves functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation improvement (updates to user guides, docstrings, or developer docs)

## Checklist:

- [X] My code follows the code style of this project and has been formatted using `black`.
- [X] I have updated the documentation (if relevant).
- [X] I have added tests that cover my changes (if relevant).
- [X] All new and existing tests passed.
- [X] `example.yaml` (and other relevant `configs/`) have been updated with new or changed options.
- [X] I have updated `CHANGELOG.md`.

### CODE REVIEW

File: nequip/model/_eng.py [hunk 1]
```diff
@@ -13,14 +14,22 @@
...
     """Base default energy model archetecture.
 
     For minimal and full configuration option listings, see ``minimal.yaml`` and ``example.yaml``.
     """
     logging.debug("Start building the network model")
 
+    builder_utils.add_avg_num_neighbors(
+        config=config, initialize=initialize, dataset=dataset
+    )
+
```
Comment (nw13slx) [inline nequip/model/_eng.py:32]: does this config revision happen after `trainer.__init__` ?

File: nequip/model/builder_utils.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+from nequip.utils import Config
+from nequip.data import AtomicDataset, AtomicDataDict
+
+
+def add_avg_num_neighbors(
+    config: Config,
+    initialize: bool,
+    dataset: Optional[AtomicDataset] = None,
+    default: Optional[Union[str, float]] = "auto",
```
Comment (nw13slx) [inline nequip/model/builder_utils.py:13]: when will the `default` argument be changed? why setting it as an argument? We don't use instantiate on this function, right?

File: nequip/model/builder_utils.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+                    "node",
+                )
+            ],
+            modes=["mean_std"],
+            stride=config.get("dataset_statistics_stride", 1),
+        )[0][0].item()
+
+    # make sure its valid
+    if ann is not None:
+        ann = float(ann)
```
Comment (nw13slx) [inline nequip/model/builder_utils.py:38]: if ann is auto and initialize is False?

File: nequip/model/_eng.py (hunk 1 above)
Comment (Linux-cpp-lisp) [reply nequip/model/_eng.py:32]: yes, is in a model builder

Comment (Linux-cpp-lisp) [reply nequip/model/_eng.py:32]: but it gets written out to the saved config.yaml:
```
avg_num_neighbors: 10.495238304138184
```

**Pull request merged**